src/Workspaces/CSharp/Portable/Utilities/SpeculationAnalyzer.cs

```diff
@@ -379,6 +379,39 @@ protected override bool ReplacementChangesSemanticsForNodeLanguageSpecific(Synta
                     }
                 }
             }
+            else if (currentOriginalNode.Kind() == SyntaxKind.CaseSwitchLabel)
+            {
+                var originalCaseSwitchLabel = (CaseSwitchLabelSyntax)currentOriginalNode;
+                var newCaseSwitchLabel = (CaseSwitchLabelSyntax)currentReplacedNode;
+
+                // If case label is changing, then need to check if the semantics will change for the switch expression.  
+                // e.g. if switch expression is "switch(x)" where "object x = 1f", then "case 1:" and "case (float) 1:" are different.
+                var originalCaseType = this.OriginalSemanticModel.GetTypeInfo(previousOriginalNode, this.CancellationToken).Type;
+                var newCaseType = this.SpeculativeSemanticModel.GetTypeInfo(previousReplacedNode, this.CancellationToken).Type;
+
+                if (originalCaseType == newCaseType)
+                    return false;
+
+                var oldSwitchStatement = originalCaseSwitchLabel?.Parent?.Parent as SwitchStatementSyntax;
+                var newSwitchStatement = newCaseSwitchLabel?.Parent?.Parent as SwitchStatementSyntax;
+                if (oldSwitchStatement == null || newSwitchStatement == null)
+                {
+                    return false;
+                }
+
+                var originalConversion = this.OriginalSemanticModel.ClassifyConversion(oldSwitchStatement.Expression, originalCaseType);
+                var newConversion = this.SpeculativeSemanticModel.ClassifyConversion(newSwitchStatement.Expression, newCaseType);
+                
+                // if conversion only exists for either original or new, then semantics changed.
+                if (originalConversion.Exists != newConversion.Exists)
+                {
+                    return true;
+                }
+
+                // Same conversion cannot result in both originalCaseType and newCaseType, which means the semantics changed
+                // (since originalCaseType != newCaseType)
+                return originalConversion == newConversion;
```

Comment:
Here, I am being quite generic and trying to check if the conversions from switch expression to `originalCaseType` and `newCaseType` are the same, then the resulting types should have been the same in order to have the same semantics. Since the types are different at this point, have the same conversions mean that the semantics actually changed. 
I couldn't think of any exceptional case that this would break something else.